### ISSUE TITLE
{"DelayPID":60}  when re-start during filter pump window

### ISSUE THREAD

**Issue by bbwouters** [Loic74650/PoolMaster#19]
Are we sure if the system is re-booted during the Flirtation pump window/ hours the PID regulation is delayed and does not start immediately regulating based on water left in the tubes (and therefore not reflecting real water quality?  Same question is valid when pump is manually stopped and started  during filtration hours. (As a matter of safety).

So does PHRegulationCallback(Task* me) checks when Firltration pump was last started? I don't think so but your insight would be appreciated.

**Comment by Loic74650** [comment]
Correct it does not. 

I'd like the PHRegulationCallback(Task* me) to remain as clean and simple as possible, so I suggest that any future checks should be either performed in the generic loop or in the lower level Pump class.

In this particular case the simplest way to do this I think, is to reset the FiltrationPump.UpTime property after a manual stop/restart of the filtration. That way it will wait for another DelayPID before it regulates again. The downside is that it will messup the filtration real duration statistic for that particular day (maybe not so critical). 

A slightly more correct/complex way of doing things would be to increase the value of storage.DelayPIDs variable by the FiltrationPump.UpTime value after each start/stop of the filtration:

storage.DelayPIDs += FiltrationPump.UpTime

That way when it restarts, it has to wait for another while before starting the regulation again. 
At midnight, storage.DelayPIDs should be reset to its default value.

We also have to discuss whether EmergencyStopFiltPump should be reset at midnight or not (ie. should the double-tap feature be used for long-term filtration management or just small daily tasks)

**Comment by bbwouters** [comment]
In my opinion: EmergencyStopFiltPump should be reset at midnight , "emergency" should the double-tap feature be used for small daily tasks, but a warning in the documentation it will restart next day. 


Van: "Loic74650" <notifications@github.com> 
Aan: "Loic74650/PoolMaster" <PoolMaster@noreply.github.com> 
Cc: "bwouters" <bwouters@telenet.be>, "Author" <author@noreply.github.com> 
Verzonden: Zondag 31 mei 2020 15:13:15 
Onderwerp: Re: [Loic74650/PoolMaster] {"DelayPID":60} when re-start during filter pump window (#19) 





Correct it does not. 

I'd like the PHRegulationCallback(Task* me) to remain as clean and simple as possible, so I suggest that any future checks should be either performed in the generic loop or in the lower level Pump class. 

In this particular case the simplest way to do this I think, is to reset the FiltrationPump.UpTime property after a manual stop/restart of the filtration. That way it will wait for another DelayPID before it regulates again. The downside is that it will messup the filtration real duration statistic for that particular day (maybe not so critical). 

A slightly more correct/complex way of doing things would be to increase the value of storage.DelayPIDs variable by the FiltrationPump.UpTime value after each start/stop of the filtration: 

storage.DelayPIDs += FiltrationPump.UpTime 

That way when it restarts, it has to wait for another while before starting the regulation again. 
At midnight, storage.DelayPIDs should be reset to its default value. 

We also have to discuss whether EmergencyStopFiltPump should be reset at midnight or not (ie. should the double-tap feature be used for long-term filtration management or just small daily tasks)

**Comment by Loic74650** [comment]
`In my opinion: EmergencyStopFiltPump should be reset at midnight , "emergency" should the double-tap feature be used for small daily tasks, but a warning in the documentation it will restart next day.`

I agree and will make the change

**Comment by bbwouters** [comment]
unsigned long LastStartTime; 

can also be used to make a check with to take into account PID delay 

I'll have a look 


Van: "Loic74650" <notifications@github.com> 
Aan: "Loic74650" <PoolMaster@noreply.github.com> 
Cc: "bwouters" <bwouters@telenet.be>, "Author" <author@noreply.github.com> 
Verzonden: Zondag 31 mei 2020 15:13:15 
Onderwerp: Re: [Loic74650/PoolMaster] {"DelayPID":60} when re-start during filter pump window (#19) 





Correct it does not. 

I'd like the PHRegulationCallback(Task* me) to remain as clean and simple as possible, so I suggest that any future checks should be either performed in the generic loop or in the lower level Pump class. 

In this particular case the simplest way to do this I think, is to reset the FiltrationPump.UpTime property after a manual stop/restart of the filtration. That way it will wait for another DelayPID before it regulates again. The downside is that it will messup the filtration real duration statistic for that particular day (maybe not so critical). 

A slightly more correct/complex way of doing things would be to increase the value of storage.DelayPIDs variable by the FiltrationPump.UpTime value after each start/stop of the filtration: 

storage.DelayPIDs += FiltrationPump.UpTime 

That way when it restarts, it has to wait for another while before starting the regulation again. 
At midnight, storage.DelayPIDs should be reset to its default value. 

We also have to discuss whether EmergencyStopFiltPump should be reset at midnight or not (ie. should the double-tap feature be used for long-term filtration management or just small daily tasks)

**Comment by Loic74650** [comment]
`unsigned long LastStartTime; can also be used to make a check with to take into account PID delay`

Correct. LastStartTime could be used to check how long the filtration was stopped for. If longer than, say, 1 hour, a delay should be applied before restarting the PIDs

**Comment by Loic74650** [comment]
I gave some thoughts to this issue and how to handle it. 
It becomes unecessarily complicated once you wish to handle several double-taps in a row. 

Therefore v5.0.0 is now using a simple check before it starts the PIDS, I have replaced the following condition:

(FiltrationPump.UpTime/1000/60 > storage.DelayPIDs)

with:

(FiltrationPump.LastStartTime / 1000 / 60 > storage.DelayPIDs)

It has the inconvenience to enforce a DelayPID every time you stop/restart the filtration but in practice that should not be abig issue and that keeps things simple

**Comment by bbwouters** [comment]
Perfect solution! 


Van: "Loic74650" <notifications@github.com> 
Aan: "Loic74650" <PoolMaster@noreply.github.com> 
Cc: "bwouters" <bwouters@telenet.be>, "Author" <author@noreply.github.com> 
Verzonden: Dinsdag 2 juni 2020 12:49:27 
Onderwerp: Re: [Loic74650/PoolMaster] {"DelayPID":60} when re-start during filter pump window (#19) 





I gave some thoughts to this issue and how to handle it. 
It becomes unecessarily complicated once you wish to handle several double-taps in a row. 

Therefore v5.0.0 is now using a simple check before it starts the PIDS, I have replaced the following condition: 

(FiltrationPump.UpTime/1000/60 > storage.DelayPIDs) 

with: 

(FiltrationPump.LastStartTime / 1000 / 60 > storage.DelayPIDs) 

It has the inconvenience to enforce a DelayPID every time you stop/restart the filtration but in practice that should not be abig issue and that keeps things simple

**Comment by Loic74650** [comment]
Hi Boris

 

Found the bug with DelayPID behaviour and the use of LastStartTime

I fixed it

 

Cheers

Loic

 

De : bbwouters <notifications@github.com> 
Envoyé : dimanche 31 mai 2020 23:08
À : Loic74650/PoolMaster <PoolMaster@noreply.github.com>
Cc : Loic74650 <loic74650@gmail.com>; Comment <comment@noreply.github.com>
Objet : Re: [Loic74650/PoolMaster] {"DelayPID":60} when re-start during filter pump window (#19)

 

unsigned long LastStartTime; 

can also be used to make a check with to take into account PID delay 

I'll have a look 


Van: "Loic74650" <notifications@github.com> 
Aan: "Loic74650" <PoolMaster@noreply.github.com> 
Cc: "bwouters" <bwouters@telenet.be>, "Author" <author@noreply.github.com> 
Verzonden: Zondag 31 mei 2020 15:13:15 
Onderwerp: Re: [Loic74650/PoolMaster] {"DelayPID":60} when re-start during filter pump window (#19) 





Correct it does not. 

I'd like the PHRegulationCallback(Task* me) to remain as clean and simple as possible, so I suggest that any future checks should be either performed in the generic loop or in the lower level Pump class. 

In this particular case the simplest way to do this I think, is to reset the FiltrationPump.UpTime property after a manual stop/restart of the filtration. That way it will wait for another DelayPID before it regulates again. The downside is that it will messup the filtration real duration statistic for that particular day (maybe not so critical). 

A slightly more correct/complex way of doing things would be to increase the value of storage.DelayPIDs variable by the FiltrationPump.UpTime value after each start/stop of the filtration: 

storage.DelayPIDs += FiltrationPump.UpTime 

That way when it restarts, it has to wait for another while before starting the regulation again. 
At midnight, storage.DelayPIDs should be reset to its default value. 

We also have to discuss whether EmergencyStopFiltPump should be reset at midnight or not (ie. should the double-tap feature be used for long-term filtration management or just small daily tasks)

**Comment by bbwouters** [comment]
Perfect! 


Van: "Loic74650" <notifications@github.com> 
Aan: "Loic74650" <PoolMaster@noreply.github.com> 
Cc: "bwouters" <bwouters@telenet.be>, "Author" <author@noreply.github.com> 
Verzonden: Dinsdag 16 juni 2020 11:17:48 
Onderwerp: Re: [Loic74650/PoolMaster] {"DelayPID":60} when re-start during filter pump window (#19) 



Hi Boris 



Found the bug with DelayPID behaviour and the use of LastStartTime 

I fixed it 



Cheers 

Loic 



De : bbwouters <notifications@github.com> 
Envoyé : dimanche 31 mai 2020 23:08 
À : Loic74650/PoolMaster <PoolMaster@noreply.github.com> 
Cc : Loic74650 <loic74650@gmail.com>; Comment <comment@noreply.github.com> 
Objet : Re: [Loic74650/PoolMaster] {"DelayPID":60} when re-start during filter pump window (#19) 



unsigned long LastStartTime; 

can also be used to make a check with to take into account PID delay 

I'll have a look 


Van: "Loic74650" <notifications@github.com> 
Aan: "Loic74650" <PoolMaster@noreply.github.com> 
Cc: "bwouters" <bwouters@telenet.be>, "Author" <author@noreply.github.com> 
Verzonden: Zondag 31 mei 2020 15:13:15 
Onderwerp: Re: [Loic74650/PoolMaster] {"DelayPID":60} when re-start during filter pump window (#19) 





Correct it does not. 

I'd like the PHRegulationCallback(Task* me) to remain as clean and simple as possible, so I suggest that any future checks should be either performed in the generic loop or in the lower level Pump class. 

In this particular case the simplest way to do this I think, is to reset the FiltrationPump.UpTime property after a manual stop/restart of the filtration. That way it will wait for another DelayPID before it regulates again. The downside is that it will messup the filtration real duration statistic for that particular day (maybe not so critical). 

A slightly more correct/complex way of doing things would be to increase the value of storage.DelayPIDs variable by the FiltrationPump.UpTime value after each start/stop of the filtration: 

storage.DelayPIDs += FiltrationPump.UpTime 

That way when it restarts, it has to wait for another while before starting the regulation again. 
At midnight, storage.DelayPIDs should be reset to its default value. 

We also have to discuss whether EmergencyStopFiltPump should be reset at midnight or not (ie. should the double-tap feature be used for long-term filtration management or just small daily tasks)
